### ISSUE TITLE
Camera2D limits don't clamp camera position

### ISSUE THREAD

**Issue by me2beats** [godotengine/godot#62441]
### Godot version

3.5rc4

### System information

windows

### Issue description

Camera2D limits don't clamp camera position

### Steps to reproduce

Add a 2D scene with a Camera2D, set limits.
move camera to reach the limits.
you can go further but position still increases.
This leads to the effect that camera is kinda sticked to the corner: you scroll back but it doesn't scroll.

It seems clamping camera position fixes it, but this should be default behavior imo.

### Minimal reproduction project

[camera_limits.zip](https://github.com/godotengine/godot/files/8989163/camera_limits.zip)

in the project camera limits work as I expect, but if you comment lines with clamp, you will see the issue

**Comment by me2beats** [comment]
```
var pos = position-event.relative*zoom
var x = zoom.x*viewport_size.x/2
var y = zoom.y*viewport_size.y/2
pos.x = clamp(pos.x, limit_left+x, limit_right-x)
pos.y = clamp(pos.y, limit_top+y, limit_bottom-y)
```
this is better

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/56336.

**Comment by me2beats** [comment]
https://github.com/godotengine/godot-proposals/issues/1640 is for offset tho

**Comment by madmiraal** [comment]
> Camera2D limits don't clamp camera position

This is correct. `Camera2D`'s limits don't clamp the camera's position. `Camera2D`'s limits set the limits of camera's boundaries i.e. the edges of the camera (not the camera's position). The `Camera2D`'s boundaries won't go beyond the limits.

**Note:** If, as in your example, the difference between the [`limit_left`](https://docs.godotengine.org/en/stable/classes/class_camera2d.html#class-camera2d-property-limit-left) and [`limit_right`](https://docs.godotengine.org/en/stable/classes/class_camera2d.html#class-camera2d-property-drag-margin-right) is less than the camera display's width or the difference between the [`limit_top`](https://docs.godotengine.org/en/stable/classes/class_camera2d.html#class-camera2d-property-limit-top) and [`limit_bottom`](https://docs.godotengine.org/en/stable/classes/class_camera2d.html#class-camera2d-property-limit-bottom) is less that the camera display's height (See [Project Settings](https://docs.godotengine.org/en/stable/classes/class_projectsettings.html) [`display/window/size/width`](https://docs.godotengine.org/en/stable/classes/class_projectsettings.html#class-projectsettings-property-display-window-size-width) and [`display/window/size/height`](https://docs.godotengine.org/en/stable/classes/class_projectsettings.html#class-projectsettings-property-display-window-size-height)), then the camera's boundaries must exceed these limits effectively leading to undefined behaviour, which is why it appears to stick to one corner. It cannot squeeze between the limits; so it does the best it can: it puts the camera's top-right corner in the top-right corner of the limits.

**Comment by me2beats** [comment]
The question is why not clamp

**Comment by madmiraal** [comment]
This is how it's designed. You define the boundary area of your game using the camera limits. The camera is free to show anything within those limits, but won't show anything beyond those limits.

**Comment by me2beats** [comment]
Why can't the camera just stop when it reaches the boundary and don't change the position anymore.
Really interesting why this is designed this way and when this would be useful instead of just clamping the position

**Comment by me2beats** [comment]
As an end user I just want the camera to work right.
But I'm facing it and I need to create the code to manually clamp the position.
Everyone will do this, because this "sticking" feels wrong.
If this is really needed to be designed this way, could there be an option like "also clamp the position!"

**Comment by madmiraal** [comment]
> Why can't the camera just stop when it reaches the boundary and don't change the position anymore.

That's exactly what is does.

> As an end user I just want the camera to work right.
But I'm facing it and I need to create the code to manually clamp the position.
Everyone will do this, because this "sticking" feels wrong.
If this is really needed to be designed this way, could there be an option like "also clamp the position!"

The limits are designed to constrain the camera display area. The problem with the example you've provided is that the camera limits are smaller than the camera display area.

This is not a bug. It's how it's how camera limits have been designed, and they work for most users. If you have a question about how camera limits work, please use one of the other forums where you question can be better answered: https://godotengine.org/community

**Comment by madmiraal** [comment]
This is done so that you can, for example, position the camera anywhere in the game e.g. the start of your next level, and then change the camera limits from surrounding the current level to surrounding your new level. With [`smoothing_enabled`](https://docs.godotengine.org/en/stable/classes/class_camera2d.html#class-camera2d-property-smoothing-enabled) and [`limit_smoothed`](https://docs.godotengine.org/en/stable/classes/class_camera2d.html#class-camera2d-property-limit-smoothed) both set to `true`, the camera will then gently scroll from the end of the current level to the start of the next level.

**Comment by Luistorch** [comment]
If I may add a possible solution for @me2beats 
```
if(dragging):
    nextCameraPositon.x = clamp(self.position.x + (mouseInitPos.x - get_global_mouse_position().x) * delta * dragSpeed * self.zoom.x, left_limit, right_limit)
    nextCameraPositon.y = clamp(self.position.y + (mouseInitPos.y - get_global_mouse_position().y) * delta * dragSpeed * self.zoom.y, top_limit, bottom_limit)
    self.position = nextCameraPositon
```
This piece inside the _process function clamps the position of the camera inside 4 numbers set as top, bottom, left and right limits. As it works every frame it does not affect the visuals

**Comment by jedomed** [comment]
I agree with me2beats that there should be an option to clamp the position by default. I just spent several hours trying to figure out why there was a delay when panning my camera from one edge of the screen to the other, only to find out that the camera's real position was completely outside the limits...
Or at least it be clearer in the [docs](https://docs.godotengine.org/en/stable/classes/class_camera2d.html)... there's the line "Camera2D node's position doesn't represent the actual position of the screen, which may differ due to applied smoothing or limits", and maybe this is a me problem but I didn't realize this was my problem even after reading it because I had no idea what "position of the screen" was supposed to mean, let alone that what it's trying to say is "what the camera is actually showing you"?

**Comment by timoschwarzer** [comment]
If you need to get the displayed camera position, use `Camera2D.get_screen_center_position()`.

**Comment by Birdulon** [comment]
> If you need to get the displayed camera position, use `Camera2D.get_screen_center_position()`.

That appears to be a frame late, so responsive limits will still need to do manual viewport math, as a naive `camera.position = camera.get_screen_center_position()` at the end of your camera motion code will just prevent the camera from moving.

**Comment by timoschwarzer** [comment]
Is it still one frame late if you set the process priority of the camera to process before whatever script is running that calls `get_screen_center_position()`?

**Comment by Birdulon** [comment]
Not sure, but using another script gave me the idea of chucking that on the camera's `process()` and it seems to work well enough regardless of priority, with another script changing its position in `process()`. So my frame late assumption seems wrong in a multi-node context.
